File: Data/Config/auto-csv/AutoCsvCode/Cpp/skill/SkillLevelConfig.cpp
#include "skill/SkillLevelConfig.h"
#include "CsvParser/csv.h"

namespace Config
{
     static const char * Field_Name_id = "id";
     static const char * Field_Name_level = "level";

    bool SkillLevelConfig::Init(std::map<std::string, std::string> kvPairs, ConfigCheckFunc func)
    {
        bool all_ok = true;
        all_ok = all_ok && kvPairs.count(Field_Name_id) > 0 && ConfigUtil::Str2BaseValue (kvPairs[Field_Name_id], id);
        all_ok = all_ok && kvPairs.count(Field_Name_level) > 0 && ConfigUtil::Str2BaseValue (kvPairs[Field_Name_level], level);
        if (all_ok && nullptr != func)
            all_ok &= func(this);
        return all_ok;
    }

    SkillLevelConfigSet::~SkillLevelConfigSet()
    {
        for (auto cfg : cfg_vec)
        {
            delete cfg;
        }
    }

    bool SkillLevelConfigSet::Load(std::string file_path)
    {
        io::CSVReader<2, io::trim_chars<' ', '\t'>, io::double_quote_escape<',', '\"'>, io::no_comment> csv_reader(file_path);
        csv_reader.read_header(io::ignore_extra_column,
            Field_Name_id,
            Field_Name_level
            );

        std::map<std::string, std::string> kvParis;
        kvParis[Field_Name_id] = "";
        kvParis[Field_Name_level] = "";

        bool all_ok = true;
        int curr_row = 0;
        while (csv_reader.read_row(
            kvParis[Field_Name_id],
            kvParis[Field_Name_level]
            ))
        {            
            if (++ curr_row <= 1)
                continue;
            if (kvParis[Field_Name_id].empty())
                continue;
            SkillLevelConfig *cfg = new SkillLevelConfig();
            all_ok &= cfg->Init(kvParis, cfg_check_fun);
            if (!all_ok)
                break;
            cfg_vec.push_back(cfg);
        }
        if (all_ok)
        {
            // gen key
            for (auto cfg : cfg_vec)
            {
                {
                    auto it = id_to_group.find(cfg->id);
                    if (id_to_group.end() == it)
                    {
                        auto ret_it = id_to_group.insert(std::make_pair(cfg->id, std::vector<SkillLevelConfig *>()));
                        if (!ret_it.second)
                        {
                            all_ok = false;
                            break;
                        }
                        it = ret_it.first;
                    }
                    it->second.push_back(cfg);
                }
            }
        }
        if (nullptr != cfg_set_check_fun)
        {
            all_ok = all_ok && cfg_set_check_fun(this);
        }
        return all_ok;
    }
}

File: Data/Config/auto-csv/AutoCsvCode/CSharp/skill/SkillLevelConfig.cs
using System;
using System.Collections.Generic;
using LumenWorks.Framework.IO.Csv;
using System.IO;

namespace Config
{
    internal class FieldName_SkillLevelConfig
    {
        public const string Field_Name_id = "id";
        public const string Field_Name_level = "level";
    }

    public class SkillLevelConfig
    {
        public int id;
        public int level;

        public delegate bool ConfigCheckFunc(SkillLevelConfig cfg);
        public bool Init(Dictionary<string, string> kvPairs, ConfigCheckFunc func)
        {
            bool all_ok = true;
        all_ok = all_ok && kvPairs.ContainsKey(FieldName_SkillLevelConfig.Field_Name_id) && ConfigUtil.Str2BaseValue (kvPairs[FieldName_SkillLevelConfig.Field_Name_id], ref id);
        all_ok = all_ok && kvPairs.ContainsKey(FieldName_SkillLevelConfig.Field_Name_level) && ConfigUtil.Str2BaseValue (kvPairs[FieldName_SkillLevelConfig.Field_Name_level], ref level);
            if (all_ok && null != func)
                all_ok &= func(this);
            return all_ok;
        }
    }

    public class SkillLevelConfigSet
    {
        public string errMsg = string.Empty;

        public SkillLevelConfig.ConfigCheckFunc cfg_check_fun = null;
        public delegate bool ConfigSetCheckFunc(SkillLevelConfigSet items);
        public ConfigSetCheckFunc cfg_set_check_fun = null;
        public List<SkillLevelConfig> cfg_vec = new List<SkillLevelConfig> ();
        public Dictionary<int, List<SkillLevelConfig>> id_to_group = new Dictionary<int, List<SkillLevelConfig>>();

        public bool Load(string file_path)
        {
            using (CsvReader csv = new CsvReader(new StreamReader(file_path), true))
            {
                csv.MissingFieldAction = MissingFieldAction.ReplaceByEmpty;
                string[] headers = csv.GetFieldHeaders();
                List<int> fieldIdxList = new List<int>();
                try
                {
                    fieldIdxList.Add(csv.GetFieldIndex(FieldName_SkillLevelConfig.Field_Name_id));
                    fieldIdxList.Add(csv.GetFieldIndex(FieldName_SkillLevelConfig.Field_Name_level));
                }
                catch (Exception e)
                {
                    errMsg = e.Message;
                    return false;
                }
                csv.ReadNextRecord();
                while (csv.ReadNextRecord())
                {
                    Dictionary<string, string> kvPairs = new Dictionary<string, string>();
                    foreach (int fieldIdx in fieldIdxList)
                    {
                        kvPairs[headers[fieldIdx]] = csv[fieldIdx];
                        
                    }
                    SkillLevelConfig cfg = new SkillLevelConfig();
                    if (!cfg.Init(kvPairs, cfg_check_fun))
                    {
                        errMsg = string.Format("SkillLevelConfig Init Fail, row {0}", csv.CurrentRecordIndex);
                        return false;
                    }
                    cfg_vec.Add(cfg);
                }
            }

            foreach (var cfg in cfg_vec)
            {
                {
                    List<SkillLevelConfig> vec = null;
                    if (!id_to_group.TryGetValue(cfg.id, out vec))
                    {
                        vec = new List<SkillLevelConfig>();
                        id_to_group[cfg.id] = vec;
                    }
                    vec.Add(cfg);
                }
            }

            if (null != cfg_set_check_fun)
            {
                if (!cfg_set_check_fun(this))
                {
                    errMsg = "SkillLevelConfigSet cfg_set_check_fun fail";
                    return false;
                }
            }

            return true;
        }
    }
}